File: services/aep-api/internal/feature/task/skills.go
// Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
//
// WSO2 LLC. licenses this file to you under the Apache License,
// Version 2.0 (the "License"); you may not use this file except
// in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package task

import (
	_ "embed"
	"sync"

	"github.com/wso2/aep/aep-api/internal/clients/agentsvc"
)

// The task-planning skill is the SINGLE SOURCE OF TRUTH at repo-root
// skills/task-planning (docs/design/agents-generation-migration.md §2). It
// cannot be go:embed-ed across the aep-api Go module boundary, so it is vendored
// here. The Task feature vendors its OWN copy (rather than reusing feature/genai's
// vendored set) because the arch allowlist forbids a task→genai import — the one
// skill the plan turn pushes is cheaper to vendor than to couple the two
// features. TestVendoredTaskPlanningSkillMatchesRepoRoot guards against drift.
// Sync command:
//
//go:generate sh -c "rm -rf assets/skills/task-planning && cp -R ../../../../../skills/task-planning assets/skills/task-planning"
//go:embed assets/skills/task-planning/SKILL.md
var taskPlanningSkillMD string

var (
	taskPlanningSkillOnce sync.Once
	taskPlanningSkill     agentsvc.Skill
)

// loadTaskPlanningSkill returns the vendored task-planning skill pushed on every
// plan turn (progressive disclosure level 1: name + description in the catalog,
// body served on demand via loadSkill).
func loadTaskPlanningSkill() agentsvc.Skill {
	taskPlanningSkillOnce.Do(func() {
		taskPlanningSkill = agentsvc.Skill{
			Name:        "task-planning",
			Description: agentsvc.FrontmatterDescription(taskPlanningSkillMD),
			Content:     taskPlanningSkillMD,
		}
	})
	return taskPlanningSkill
}

File: playground/src/kit/skills.ts
/**
 * Copyright (c) 2026, WSO2 LLC. (https://www.wso2.com).
 *
 * WSO2 LLC. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/**
 * The eval's skill resolver (ADR-0002/§12). Skills never travel in the turn
 * payload; the evals/playground resolve the repo-root library here and
 * MATERIALIZE it into the fixture mount's `_skills` snapshot
 * (`EvalWorkspace.materializeSkills`), which the service reads like any caller's.
 * This module reads repo-root `skills/<id>/SKILL.md`, parses the frontmatter for
 * `name`+`description`, and takes the body as `content`. Only `evals/` touches
 * the fs.
 */

import { existsSync, readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { parse as parseYaml } from "yaml";
// Reuse the bundle's single frontmatter grammar + LF canonicalizer so the
// SKILL.md fence parsing can't drift from the spec-file fence parsing.
import { FRONTMATTER_RE, lf } from "@aep/agent-stream";

/**
 * One resolved repo skill — the eval-side shape `materializeSkills` renders into
 * a fixture `_skills` snapshot (flat `skills/<name>/SKILL.md` + references).
 */
export interface RepoSkill {
  /** Stable id; the snapshot dir name and what the catalog lists. */
  name: string;
  /** One-line summary rendered into the SKILL.md frontmatter. */
  description: string;
  /** The full guidance body (frontmatter stripped). */
  content: string;
  /** Optional `references/<file>.md` → body (agentskills.io structure). */
  references?: Record<string, string>;
  /**
   * The authored `metadata:` block, carried through verbatim. The service keys
   * real behaviour off it — `metadata.aep.kind` (ownership) and
   * `metadata.aep.audience` (which agent may load the skill) — so dropping it
   * would materialize a library that behaves unlike any real org's: every skill
   * unmarked, so every skill loadable by every agent.
   */
  metadata?: Record<string, unknown>;
}

/** Split a `SKILL.md` into its YAML frontmatter block and markdown body. */
function splitFrontmatter(raw: string): { frontmatter: string; body: string } {
  const text = lf(raw);
  const m = FRONTMATTER_RE.exec(text);
  if (!m) return { frontmatter: "", body: text };
  return { frontmatter: m[1] ?? "", body: text.slice(m[0].length) };
}

/** Parse one `SKILL.md` into a `RepoSkill`; `name` falls back to the directory id. */
export function parseSkill(dirId: string, raw: string): RepoSkill {
  const { frontmatter, body } = splitFrontmatter(raw);
  let fm: Record<string, unknown> = {};
  if (frontmatter.trim() !== "") {
    const parsed = parseYaml(frontmatter) as unknown;
    if (parsed && typeof parsed === "object") fm = parsed as Record<string, unknown>;
  }
  const name = typeof fm.name === "string" && fm.name.trim() !== "" ? fm.name : dirId;
  const description = typeof fm.description === "string" ? fm.description : "";
  const metadata =
    fm.metadata !== null && typeof fm.metadata === "object" && !Array.isArray(fm.metadata)
      ? (fm.metadata as Record<string, unknown>)
      : undefined;
  return { name, description, content: body.trim(), ...(metadata ? { metadata } : {}) };
}

/**
 * Skill names to materialize as DISABLED, from `AEP_DISABLED_SKILLS` (comma-
 * separated). In the platform an org admin toggles availability and the flag
 * lands in the org repo's `skills-manifest.json`; the playground has no admin
 * and no org repo, so this env var stands in for that toggle — enough to
 * exercise the read side (a disabled skill must vanish from the turn's catalog
 * entirely, not merely become unloadable).
 */
export function disabledSkillNames(): string[] {
  return (process.env.AEP_DISABLED_SKILLS ?? "")
    .split(",")
    .map((s) => s.trim())
    .filter((s) => s !== "");
}

/**
 * Load the whole skill library from `skillsDir` (repo-root `skills/`): every
 * immediate subdirectory holding a `SKILL.md`, sorted by id. Returns `[]` when
 * the directory is absent, so a checkout without `skills/` runs skill-free.
 */
export function loadRepoSkills(skillsDir: string): RepoSkill[] {
  if (!existsSync(skillsDir)) return [];
  return readdirSync(skillsDir, { withFileTypes: true })
    .filter((d) => d.isDirectory())
    .sort((a, b) => a.name.localeCompare(b.name))
    .flatMap((d) => {
      try {
        const skill = parseSkill(d.name, readFileSync(join(skillsDir, d.name, "SKILL.md"), "utf8"));
        const references = loadReferences(join(skillsDir, d.name));
        return [references ? { ...skill, references } : skill];
      } catch {
        return []; // a directory without a readable SKILL.md is simply not a skill
      }
    });
}

/**
 * Read `references/*.md` for one skill dir into a path→body map keyed
 * `references/<file>` (the wire shape `loadSkillReference` addresses).
 * Returns undefined — not an empty map — when the dir is absent or empty, so a
 * references-free skill stays byte-identical.
 */
function loadReferences(skillDir: string): Record<string, string> | undefined {
  const refsDir = join(skillDir, "references");
  if (!existsSync(refsDir)) return undefined;
  const out: Record<string, string> = {};
  for (const f of readdirSync(refsDir, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name))) {
    if (!f.isFile() || !f.name.endsWith(".md")) continue;
    out[`references/${f.name}`] = readFileSync(join(refsDir, f.name), "utf8");
  }
  return Object.keys(out).length > 0 ? out : undefined;
}
